src/main/java/com/exadel/discountwebapp/filter/SpecificationBuilder.java

```diff
@@ -0,0 +1,58 @@
+package com.exadel.discountwebapp.filter;
+
+import org.springframework.data.jpa.domain.Specification;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class SpecificationBuilder<T> {
+
+    private final List<SearchCriteria> criteria = new ArrayList<>();
+
+    public Specification<T> fromQuery(String query) {
+        if (query == null || query.trim().length() == 0) {
+            return null;
+        }
+
+        String regexp = "(\\w+\\.?\\w+)(:|<|>|\\*:|:\\*|\\*:\\*)([^(\\*).]+?);";
+        Pattern pattern = Pattern.compile(regexp);
```

Comment:
Let's move 'regexp' variable to a constant. And prematurely compile this regexp in a static variable just once. This approach is considered a best practice that helps to improve performance.

Reply:
done